fem/nonlininteg.cpp

```diff
@@ -673,4 +698,901 @@ void IncompressibleNeoHookeanIntegrator::AssembleElementGrad(
+            MFEM_SYNC_THREAD;
+            MFEM_FOREACH_THREAD(qz,z,Q1D)
+            {
+               MFEM_FOREACH_THREAD(qy,y,Q1D)
+               {
+                  MFEM_FOREACH_THREAD(qx,x,Q1D)
+                  {
+                     const int q = qx + (qy + qz * Q1D) * Q1D;
+                     const double z = U[1][qz][qy][qx];
+                     const double gX = QQQ0[qz][qy][qx];
+                     const double gY = QQQ1[qz][qy][qx];
+                     const double gZ = QQQ2[qz][qy][qx];
+                     const double d = gX*D(q,0,c,e) + gY*D(q,1,c,e) + gZ*D(q,2,c,e);
+                     Z[qz][qy][qx] += z * d;
+                  }
+               }
+            }
+            MFEM_SYNC_THREAD;
+         } // for each conv component
+         if (tidz == 0)
+         {
+            MFEM_FOREACH_THREAD(d,y,D1D)
+            {
+               MFEM_FOREACH_THREAD(q,x,Q1D)
+               {
+                  Bt[d][q] = b(q,d);
+               }
+            }
+         }
+         MFEM_SYNC_THREAD;
+         MFEM_FOREACH_THREAD(qz,z,Q1D)
+         {
+            MFEM_FOREACH_THREAD(qy,y,Q1D)
+            {
+               MFEM_FOREACH_THREAD(dx,x,D1D)
+               {
+                  double u = 0.0;
+                  for (int qx = 0; qx < Q1D; ++qx)
+                  {
+                     u += Z[qz][qy][qx] * Bt[dx][qx];
+                  }
+                  QQD0[qz][qy][dx] = u;
+               }
+            }
+         }
+         MFEM_SYNC_THREAD;
+         MFEM_FOREACH_THREAD(qz,z,Q1D)
+         {
+            MFEM_FOREACH_THREAD(dy,y,D1D)
+            {
+               MFEM_FOREACH_THREAD(dx,x,D1D)
+               {
+                  double u = 0.0;
+                  for (int qy = 0; qy < Q1D; ++qy)
+                  {
+                     u += QQD0[qz][qy][dx] * Bt[dy][qy];
+                  }
+                  QDD0[qz][dy][dx] = u;
+               }
+            }
+         }
+         MFEM_SYNC_THREAD;
+         MFEM_FOREACH_THREAD(dz,z,D1D)
+         {
+            MFEM_FOREACH_THREAD(dy,y,D1D)
+            {
+               MFEM_FOREACH_THREAD(dx,x,D1D)
+               {
+                  double u = 0.0;
+                  for (int qz = 0; qz < Q1D; ++qz)
+                  {
+                     u += QDD0[qz][dy][dx] * Bt[dz][qz];
+                  }
+                  Y(dx,dy,dz,cy,e) += u;
+               }
+            }
+         }
+         MFEM_SYNC_THREAD;
+      }
+   });
+}
+
+void VectorConvectionNLFIntegrator::MultPA(const Vector &x, Vector &y) const
+{
+   const int NE = ne;
+   const int D1D = maps->ndof;
+   const int Q1D = maps->nqpt;
+   const Vector &Q = pa_data;
+   const Array<double> &B = maps->B;
+   const Array<double> &G = maps->G;
+   const Array<double> &Bt = maps->Bt;
+   if (dim == 2)
+   {
+      return PAConvectionNLApply2D(NE,B,G,Bt,Q,x,y,D1D,Q1D);
+   }
+   if (dim == 3)
+   {
+      return SmemPAConvectionNLApply3D(NE,B,G,Q,x,y,D1D,Q1D);
+   }
```

Comment:
As in the case of #1164, we need to figure out a system for dealing with the template instantiation for these kernels.